smcontroller/smcontroller.go

```diff
@@ -497,52 +344,85 @@ func (controller *Controller) GetUpdateInstancesStatusChannel() <-chan []cloudpr
 }
 
 /***********************************************************************************************************************
  * Private
  **********************************************************************************************************************/
 
-func (controller *Controller) waitAndLock() {
-	controller.readyWG.Wait()
-	controller.Lock()
-}
+func (controller *Controller) processInternallMessages(ctx context.Context) {
+	for {
+		select {
+		case <-controller.connectionTimer.C:
+			log.Error("SM clients connection timeout")
 
-func (controller *Controller) connectClient(ctx context.Context, smConfig config.SMConfig, secureOpt grpc.DialOption) {
-	defer controller.readyWG.Done()
+		case internalMsg := <-controller.eventChannel:
+			log.Debug("Internal Event ", internalMsg.requestType)
 
-	connectChannel := make(chan struct{}, 1)
+			switch internalMsg.requestType {
+			case openConnection:
+				controller.handleNewConnection(internalMsg.nodeID, internalMsg.handler)
 
-	controller.clientsWG.Add(1)
+			case closeConnection:
+				controller.handleCloseConnection(internalMsg.nodeID)
 
-	go func() {
-		defer controller.clientsWG.Done()
+			default:
+				log.Error("Unsupported internal message ", internalMsg.requestType)
+			}
 
-		client, err := newSMClient(ctx, smConfig, controller.messageSender, controller.alertSender,
-			controller.monitoringSender, controller.updateInstancesStatusChan, controller.runInstancesStatusChan, secureOpt)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
 
-		connectChannel <- struct{}{}
+func (controller *Controller) handleNewConnection(nodeID string, newHandler *smHandler) {
+	controller.Lock()
+	defer controller.Unlock()
 
-		if err != nil {
-			if !strings.Contains(err.Error(), context.Canceled.Error()) {
-				log.WithField("id", smConfig.SMID).Errorf("Can't connect to SM: %s", err)
-			}
+	if handler, ok := controller.nodes[nodeID]; ok {
+		if handler != nil {
+			log.Warnf("Connection for nodeID %s already exist. Replace it", nodeID)
+		}
+	} else {
+		log.Warnf("Brand new nodeID connection with nodeID %s", nodeID)
+	}
+
+	controller.nodes[nodeID] = newHandler
 
+	for _, node := range controller.nodes {
+		if node == nil {
 			return
 		}
+	}
 
-		controller.Lock()
-		defer controller.Unlock()
+	log.Debug("All SM client connected")
+	controller.connectionTimer.Stop()
+}
 
-		controller.clients[smConfig.SMID] = client
-	}()
+func (controller *Controller) handleCloseConnection(nodeID string) {
+	controller.Lock()
+	defer controller.Unlock()
 
-	select {
-	case <-connectChannel:
+	if _, ok := controller.nodes[nodeID]; !ok {
+		log.Errorf("Connection for nodeID %s doesn't exist.", nodeID)
 
-	case <-time.After(connectClientTimeout):
-		log.WithField("id", smConfig.SMID).Error("SM connection timeout")
+		return
 	}
+
+	controller.nodes[nodeID] = nil
+}
+
+func (controller *Controller) getNodeHandlerByID(nodeID string) (*smHandler, error) {
+	if handler, ok := controller.nodes[nodeID]; ok {
```

Comment:
```go
... !ok {
    return err
}
```

Reply:
we have to return an error when the handler doesn't exist or it is nil, to avoid returning two errors. decided to use this approach.

Reply:
I didn't get your point:
```go
	if handler, ok := controller.nodes[nodeID]; !ok {
              return nil, aoserrors.Errorf("unknown nodeID %s", nodeID)
        }
	
        if handler == nil {
        	return nil, aoserrors.Errorf("connection for nodeID %s doesn't exist", nodeID)
	}

        return handler, nil
```
Same amount of returned errors. But less nested coditions,